source/Halibut/Transport/Protocol/MessageSerializer.cs

```diff
@@ -61,17 +67,28 @@ public void WriteMessage<T>(Stream stream, T message)
                 createSerializer().Serialize(bson, new MessageEnvelope<object> { Message = message });
             }
 
-            observer.MessageWritten(compressedByteCountingStream.BytesWritten);
+            observer.MessageWritten(compressedByteCountingStream.BytesWritten, 0);
         }
 
         public async Task WriteMessageAsync<T>(Stream stream, T message, CancellationToken cancellationToken)
         {
-            // TODO - ASYNC ME UP!
-            await Task.CompletedTask;
+            using var compressedByteCountingStream = new ByteCountingStream(stream, OnDispose.LeaveInputStreamOpen);
+            using var compressedInMemoryBuffer = new WriteIntoMemoryBufferStream(compressedByteCountingStream, writeIntoMemoryLimitBytes, OnDispose.LeaveInputStreamOpen);
 
-            WriteMessage(stream, message);
-        }
+            using (var zip = new DeflateStream(compressedInMemoryBuffer, CompressionMode.Compress, true))
+            using (var bson = new BsonDataWriter(zip) { CloseOutput = false })
+            {
+                // for the moment this MUST be object so that the $type property is included
+                // If it is not, then an old receiver (eg, old tentacle) will not be able to understand messages from a new sender (server)
+                // Once ALL sources and targets are deserializing to MessageEnvelope<T>, (ReadBsonMessage) then this can be changed to T
```

Comment:
This comment I am sure is misleading, we send a MessageEnvelope type over the wire. So changing that would break compatibility. I know its pre-existing but it could just be deleted.

Reply:
I don't have much context on the validity of this comment (hence leaving it 😁 ).

If you're sure we can delete, I'm happy to do so.

Delete it?